File: Python/BOJ/시뮬레이션/14503.py
//
//  14503.py
//  Algorithm
//
//  Created by 유정주 on 2023/06/04.
//

import sys; readLine = sys.stdin.readline

# 북서남동
dx = [-1, 0, 1, 0]; dy = [0, 1, 0, -1]

n, m = map(int, readLine().split())
robot = list(map(int, readLine().split()))
board = [list(map(int, readLine().split())) for _ in range(n)]

result = 0
while True:
  r, c, d = robot[0], robot[1], robot[2]

  # 1번
  if board[r][c] == 0:
    board[r][c] = -1
    result += 1

  canClean = False
  for i in reversed(range(4)):
    nd = (d + i) % 4
    nx = r + dx[nd]
    ny = c + dy[nd]

    # 3번
    if board[nx][ny] == 0:
      canClean = True
      robot = [nx, ny, nd]
      break

  # 2번
  if not canClean:
    # 후진
    nx = r - dx[d]
    ny = c - dy[d]

    if board[nx][ny] == 1:
      break
    else:
      robot = [nx, ny, d]

print(result)

File: Swift/BOJ/14000/14503.swift
//
//  14503.swift
//  SwiftAlgorithm
//
//  Created by 유정주 on 2023/08/09.
//

import Foundation

import Foundation

func solution() {
    map[robot.r][robot.c] = -1
    result += 1
    
    while true {
        if spin() {
            map[robot.r][robot.c] = -1
            result += 1
        } else {
            let backX = robot.r - dx[robot.d]
            let backY = robot.c - dy[robot.d]
            if map[backX][backY] == 1 {
                return
            } else {
                robot = (backX, backY, robot.d)
            }
        }
    }
}

func spin() -> Bool {
    var nd = robot.d
    for _ in 0..<4 {
        nd -= 1
        if nd < 0 {
            nd += 4
        }
        
        let nx = robot.r + dx[nd]
        let ny = robot.c + dy[nd]
        
        guard 0..<n ~= nx && 0..<m ~= ny && map[nx][ny] == 0 else {
            continue
        }
        
        //회전
        robot = (nx, ny, nd)
        return true
    }
    
    return false
}

//북 동 남 서
let dx = [-1, 0, 1, 0]
let dy = [0, 1, 0, -1]

let nm = readLine()!.split { $0 == " " }.map { Int(String($0))! }
let n = nm[0], m = nm[1]

let rcd = readLine()!.split { $0 == " " }.map { Int(String($0))! }
var robot: (r: Int, c: Int, d: Int) = (rcd[0], rcd[1], rcd[2])

//0: 청소 안 됨, 1: 벽이 있음
var map: [[Int]] = []
for _ in 0..<n {
    let input = readLine()!.split { $0 == " " }.map { Int(String($0))! }
    map.append(input)
}

var result = 0
solution()
print(result)
